File: src/services/PlaidService.ts
import * as SecureStore from 'expo-secure-store';
import { Transaction } from '../models/Transaction';

export class PlaidService {
  private baseURL: string;
  private accessToken: string | null = null;
  private accountsMap: Map<string, string> = new Map();

  constructor() {
    this.baseURL = process.env.EXPO_PUBLIC_BACKEND_URL || 'https://your-app.vercel.app/api';
    this.loadAccessToken();
  }

  private async loadAccessToken() {
    try {
      this.accessToken = await SecureStore.getItemAsync('plaid_access_token');
    } catch (error) {
      console.error('Failed to load access token:', error);
    }
  }

  async createLinkToken(userId?: string): Promise<string> {
    try {
      const response = await fetch(`${this.baseURL}/create-link-token`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          user_id: userId || 'default-user'
        }),
      });

      if (!response.ok) {
        throw new Error(`Failed to create link token: ${response.statusText}`);
      }

      const data = await response.json();
      return data.link_token;
    } catch (error) {
      console.error('Error creating link token:', error);
      throw error;
    }
  }

  async exchangePublicToken(publicToken: string): Promise<void> {
    try {
      const response = await fetch(`${this.baseURL}/exchange-token`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          public_token: publicToken
        }),
      });

      if (!response.ok) {
        throw new Error(`Failed to exchange token: ${response.statusText}`);
      }

      const data = await response.json();
      this.accessToken = data.access_token;

      await SecureStore.setItemAsync('plaid_access_token', data.access_token);
      await SecureStore.setItemAsync('plaid_item_id', data.item_id);
    } catch (error) {
      console.error('Error exchanging public token:', error);
      throw error;
    }
  }

  async isConnected(): Promise<boolean> {
    if (!this.accessToken) {
      await this.loadAccessToken();
    }
    return !!this.accessToken;
  }

  async disconnect(): Promise<void> {
    try {
      await SecureStore.deleteItemAsync('plaid_access_token');
      await SecureStore.deleteItemAsync('plaid_item_id');
      this.accessToken = null;
      this.accountsMap.clear();
    } catch (error) {
      console.error('Error disconnecting Plaid:', error);
    }
  }

  private async fetchTransactionsForDateRange(startDate: Date, endDate: Date): Promise<Transaction[]> {
    if (!this.accessToken) {
      throw new Error('No access token available');
    }

    try {
      const response = await fetch(`${this.baseURL}/get-transactions`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          access_token: this.accessToken,
          start_date: startDate.toISOString().split('T')[0],
          end_date: endDate.toISOString().split('T')[0],
        }),
      });

      if (!response.ok) {
        throw new Error(`Failed to fetch transactions: ${response.statusText}`);
      }

      const data = await response.json();

      if (data.accounts) {
        data.accounts.forEach((account: any) => {
          this.accountsMap.set(account.account_id, account.name || account.official_name || 'Unknown');
        });
      }

      return data.transactions.map((t: any) => ({
        ...t,
        account_display_name: this.accountsMap.get(t.account_id) || 'Unknown Account',
      }));
    } catch (error) {
      console.error('Error fetching transactions:', error);
      throw error;
    }
  }

  async fetchDailyTransactions(): Promise<Transaction[]> {
    const today = new Date();
    today.setHours(0, 0, 0, 0);
    const tomorrow = new Date(today);
    tomorrow.setDate(tomorrow.getDate() + 1);

    return this.fetchTransactionsForDateRange(today, tomorrow);
  }

  async fetchWeeklyTransactions(): Promise<Transaction[]> {
    const today = new Date();
    today.setHours(0, 0, 0, 0);

    const weekStart = new Date(today);
    weekStart.setDate(today.getDate() - today.getDay());

    const weekEnd = new Date(weekStart);
    weekEnd.setDate(weekStart.getDate() + 7);

    return this.fetchTransactionsForDateRange(weekStart, weekEnd);
  }

  async fetchMonthlyTransactions(): Promise<Transaction[]> {
    const today = new Date();
    const monthStart = new Date(today.getFullYear(), today.getMonth(), 1);
    const monthEnd = new Date(today.getFullYear(), today.getMonth() + 1, 0);

    return this.fetchTransactionsForDateRange(monthStart, monthEnd);
  }

  async fetchYearlyTransactions(): Promise<Transaction[]> {
    const today = new Date();
    const yearStart = new Date(today.getFullYear(), 0, 1);
    const yearEnd = new Date(today.getFullYear(), 11, 31);

    return this.fetchTransactionsForDateRange(yearStart, yearEnd);
  }

  async fetchYearlyTotal(): Promise<number> {
    const transactions = await this.fetchYearlyTransactions();
    return transactions.reduce((sum, t) => {
      const amount = parseFloat(t.amount.toString());
      return sum + amount;
    }, 0);
  }

  async fetchLastWeekTotal(): Promise<number> {
    const today = new Date();
    today.setHours(0, 0, 0, 0);

    const lastWeekEnd = new Date(today);
    lastWeekEnd.setDate(today.getDate() - today.getDay() - 1);

    const lastWeekStart = new Date(lastWeekEnd);
    lastWeekStart.setDate(lastWeekEnd.getDate() - 6);

    const transactions = await this.fetchTransactionsForDateRange(lastWeekStart, lastWeekEnd);
    return transactions.reduce((sum, t) => {
      const amount = parseFloat(t.amount.toString());
      return sum + amount;
    }, 0);
  }
}

File: PerchIOS/PerchIOS/PerchIOS/PerchIOS/Services/PlaidService.swift
import Foundation
import LinkKit

@MainActor
class PlaidService: ObservableObject {
    @Published var isConnected = false
    @Published var isLoading = false
    @Published var errorMessage: String?

    private let keychain = KeychainManager.shared
    private var linkHandler: Handler?
    private let backendURL: String

    private let accessTokenKey = "plaid_access_token"
    private let itemIdKey = "plaid_item_id"

    init() {
        self.backendURL = Configuration.dynamicBackendURL
        checkConnectionStatus()
    }

    func checkConnectionStatus() {
        isConnected = keychain.get(accessTokenKey) != nil
    }

    func createLinkToken(userId: String = "default-user") async throws -> String {
        guard let url = URL(string: "\(backendURL)/create-link-token") else {
            throw PlaidError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body = ["user_id": userId]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw PlaidError.networkError
        }

        let result = try JSONDecoder().decode(LinkTokenResponse.self, from: data)
        return result.linkToken
    }

    func exchangePublicToken(_ publicToken: String) async throws {
        guard let url = URL(string: "\(backendURL)/exchange-token") else {
            throw PlaidError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body = ["public_token": publicToken]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw PlaidError.networkError
        }

        let result = try JSONDecoder().decode(ExchangeTokenResponse.self, from: data)

        // Store tokens securely
        keychain.save(result.accessToken, forKey: accessTokenKey)
        keychain.save(result.itemId, forKey: itemIdKey)

        isConnected = true
    }

    func presentPlaidLink() async {
        isLoading = true
        errorMessage = nil

        do {
            let linkToken = try await createLinkToken()

            var configuration = LinkTokenConfiguration(
                token: linkToken,
                onSuccess: { [weak self] success in
                    Task { @MainActor in
                        do {
                            try await self?.exchangePublicToken(success.publicToken)
                            self?.isLoading = false
                        } catch {
                            self?.errorMessage = error.localizedDescription
                            self?.isLoading = false
                        }
                    }
                }
            )

            configuration.onExit = { [weak self] exit in
                if let error = exit.error {
                    self?.errorMessage = error.localizedDescription
                }
                self?.isLoading = false
            }

            let result = Plaid.create(configuration)
            switch result {
            case .success(let handler):
                self.linkHandler = handler
                handler.open(presentUsing: .viewController(UIApplication.shared.windows.first?.rootViewController))
            case .failure(let error):
                self.errorMessage = error.localizedDescription
                self.isLoading = false
            }

        } catch {
            self.errorMessage = error.localizedDescription
            self.isLoading = false
        }
    }

    func disconnect() {
        keychain.delete(accessTokenKey)
        keychain.delete(itemIdKey)
        isConnected = false
    }

    func fetchTransactions(startDate: Date, endDate: Date) async throws -> [Transaction] {
        guard let accessToken = keychain.get(accessTokenKey) else {
            throw PlaidError.noAccessToken
        }

        guard let url = URL(string: "\(backendURL)/get-transactions") else {
            throw PlaidError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyy-MM-dd"

        let body: [String: Any] = [
            "access_token": accessToken,
            "start_date": dateFormatter.string(from: startDate),
            "end_date": dateFormatter.string(from: endDate)
        ]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw PlaidError.networkError
        }

        let result = try JSONDecoder().decode(TransactionsResponse.self, from: data)

        // Map account names
        var accountsMap: [String: String] = [:]
        for account in result.accounts {
            accountsMap[account.accountId] = account.name ?? account.officialName ?? "Unknown"
        }

        // Map Plaid transactions to our Transaction model
        return result.transactions.map { plaidTransaction in
            Transaction(
                id: Int(plaidTransaction.transactionId.hashValue),
                date: plaidTransaction.date,
                payee: plaidTransaction.name,
                amount: String(plaidTransaction.amount),
                currency: plaidTransaction.isoCurrencyCode ?? "usd",
                plaidAccountName: accountsMap[plaidTransaction.accountId],
                plaid_account_name: nil,
                assetName: nil,
                asset_name: nil,
                account_display_name: accountsMap[plaidTransaction.accountId],
                excludeFromBudget: nil,
                exclude_from_budget: nil,
                excludeFromTotals: false,
                exclude_from_totals: nil,
                isGroup: nil,
                is_group: nil,
                groupId: nil,
                group_id: nil,
                categoryName: plaidTransaction.category?.first,
                category_name: nil
            )
        }
    }
}

// MARK: - Plaid Error Types
enum PlaidError: LocalizedError {
    case invalidURL
    case networkError
    case noAccessToken

    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Invalid backend URL"
        case .networkError:
            return "Network error occurred"
        case .noAccessToken:
            return "No Plaid access token found"
        }
    }
}

// MARK: - Response Models
struct LinkTokenResponse: Codable {
    let linkToken: String

    enum CodingKeys: String, CodingKey {
        case linkToken = "link_token"
    }
}

struct ExchangeTokenResponse: Codable {
    let accessToken: String
    let itemId: String

    enum CodingKeys: String, CodingKey {
        case accessToken = "access_token"
        case itemId = "item_id"
    }
}

struct TransactionsResponse: Codable {
    let accounts: [PlaidAccount]
    let transactions: [PlaidTransaction]
}

struct PlaidAccount: Codable {
    let accountId: String
    let name: String?
    let officialName: String?

    enum CodingKeys: String, CodingKey {
        case accountId = "account_id"
        case name
        case officialName = "official_name"
    }
}

struct PlaidTransaction: Codable {
    let transactionId: String
    let accountId: String
    let amount: Double
    let date: String
    let name: String
    let category: [String]?
    let isoCurrencyCode: String?

    enum CodingKeys: String, CodingKey {
        case transactionId = "transaction_id"
        case accountId = "account_id"
        case amount
        case date
        case name
        case category
        case isoCurrencyCode = "iso_currency_code"
    }
}